### PR TITLE
ath79: add support for GL.iNet GL-USB150

### PR DESCRIPTION
**Pull request by ptpt52**
Add support for the ar71xx supported GL.iNet GL-USB150 to ath79.

GL.iNet GL-USB150 is an USB dongle WiFi router, based on Atheros AR9331.

Specification:

- 400/400/200 MHz (CPU/DDR/AHB)
- 64 MB of RAM (DDR2)
- 16 MB of FLASH (SPI NOR)
- Realtek RTL8152B USB to Ethernet bridge (connected with AR9331 PHY4)
- 1T1R 2.4 GHz
- 2x LED, 1x button
- UART header on PCB

Flash instruction:

Vendor software is based on openwrt so you can flash the sysupgrade
image via the vendor GUI or using command line sysupgrade utility.
Make sure to not save configuration over reflash as uci settings
differ between versions.

### CODE REVIEW

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+	gpio-export {
+		compatible = "gpio-export";
+
+		gpio_lan_reset {
+			gpio-export,name = "lan:reset";
+			gpio-export,output = <1>;
+			gpios = <&gpio 7 GPIO_ACTIVE_LOW>;
+		};
+	};
```
Comment (ptpt52) [inline target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:56]: for the `gpio_lan_reset ` 
I am not quite sure about this

in ar71xxx:
```
    gpio_request_one(GL_USB150_GPIO_LAN_RESET,
             GPIOF_OUT_INIT_LOW | GPIOF_EXPORT_DIR_FIXED,
             "LAN reset");
```

Comment (ptpt52) [inline target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:56]: for the `gpio_lan_reset ` 
I am not quite sure about this

in ar71xxx:
```
    gpio_request_one(GL_USB150_GPIO_LAN_RESET,
             GPIOF_OUT_INIT_LOW | GPIOF_EXPORT_DIR_FIXED,
             "LAN reset");
```

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+};
```
Comment (ptpt52) [inline target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:57]: for the `gpio_lan_reset ` 
I am not quite sure about this

in ar71xxx:
```
    gpio_request_one(GL_USB150_GPIO_LAN_RESET,
             GPIOF_OUT_INIT_LOW | GPIOF_EXPORT_DIR_FIXED,
             "LAN reset");
```

Comment (s-2) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:58]: If that pin is required to be pulled low for the USB-Ethernet Transceiver to work, it could probably be declared as a `reset-gpio` for the corresponding phy node (according to the bootlog this is `phy4` of `mdio1`), although it is technically not the phy itself that would react upon a reset; so, I'm not sure whether this would be acceptable anyways:

```
&swphy4 {
	reset-gpios = <&gpio 7 GPIO_ACTIVE_HIGH>;
}
```
It is active high here, since the driver will write the reset output to logic 0 ("not reset") for normal operation, resulting in a low level on the line (as it is currently achieved by writing 1 to an active low gpio).

Comment (ptpt52) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:58]: @s-2 thanks, but I never seem this
and should this be a  regulator?
```
    eth_vbus {
        compatible = "regulator-fixed";
        regulator-name = "eth_vbus";
        regulator-min-microvolt = <3300000>;
        regulator-max-microvolt = <3300000>;
        regulator-always-on;
        gpio = <&gpio 7 GPIO_ACTIVE_LOW>;
    };
```

Comment (s-2) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:58]: That would be another way to declare it, and probably matches the purpose of the pin even more closely. 

It is actually still an issue that certain devices require special pins to be set, while OpenWRT is trying to abandon the upstream-incompatible `gpio-export` (c.f. https://github.com/openwrt/openwrt/pull/1366#issuecomment-424834448).

However there seems to be no perfect replacement for all use cases yet, e.g. on DIR-510L we used `ucidef_add_gpio_switch` to enable USB power (as recommended in the PR), which unfortunately broke extroot (c.f. https://github.com/openwrt/openwrt/pull/3191#issuecomment-657658654).
Just like `gpio-export`, it allows the user change the pin state during runtime, but for this router it would probably break failsafe mode via USB.

In conclusion, I don't have any opinion on whether to use `reset-gpios`, `regulator-fixed` or `gpio-hog` (or deprecated `gpio-export`, allowing the user to disable USB), so let's wait to hear comments from maintainers.

Comment (ptpt52) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:58]: tested and yes it is GPIO_ACTIVE_HIGH
will change in next push.

**Comment by ptpt52** [comment]
test resault to set on the gpio 7:
```
echo "0" > /sys/class/gpio/lan\:reset/value
link down

echo "0" > /sys/class/gpio/lan\:reset/value
link up
```

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts [hunk 3]
```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "ar9331.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+
```
Comment (adschm) [inline target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:8]: remove second empty line.

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts (hunk 2 above)
Comment (adschm) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts]: IIRC that last state was the author might choose between gpio-hog, gpio-export and user-space ucidef depending on what he/she preferred and what seemed reasonable for the device at hand.
I can comment on the benefits or drawbacks of regulator-fixed, however it seems like that one is used for the other glinet devices with ar9331 in ath79 target so far. ar9330 GL-AR150 uses a gpios property at &usb_phy instead.

So, it looks like there is room to choose.

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+&eth1 {
+	status = "okay";
+
+	compatible = "syscon", "simple-mfd";
+};
+
+&spi {
+	status = "okay";
+
+	num-chipselects = <1>;
```
Comment (adschm) [inline target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:81]: Can somebody enlighten me on num-cs vs. num-chipselects?

**Comment by adschm** [comment]
Just had a look at ar71xx files, this looks generally fine.

**Comment by ptpt52** [comment]
I found an issue: sysupgrade never success

don't know if it is related to low memory(64M device)

but I test on another ath79 device, which has 128M memory, no sysupgrade issue.

how to do debug on this? I cannot access TTL on this gl-usb150

Who has a 64m memory device, maybe you can test sysupgrade?

**Comment by ptpt52** [comment]
I have to do mtd write to upgrade the firmware, sysupgrade never success
it just reboot when I do sysupgrade.

**Comment by ptpt52** [comment]
what a strange, tested on Phicomm K2T, which is 64M memory, no sysupgrade issue.

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts (hunk 4 above)
Comment (981213) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:80]: A quick grep in kernel drivers/spi shows that there's no such a property called num-chipselects.
num-cs is the property to use if a driver wants to specify number of available chipselects in dts.
both spi drivers in ath79 use fixed chipselect count, so num-cs can be omitted.

Comment (adschm) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:80]: That's interesting. We have about 20 num-chipselects in ath79 and ipq40xx, and actually 171 num-cs in ath79. So, you mean we could remove these for SPI-only devices?

**Comment by 1Jeff1** [comment]
I tried sysupgrade and the upgrade process never started, it reboot immediately.

```
181.807306] wlan0: associated
[  181.826880] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[  935.886127] reboot: Re��

*********************************************
*        U-Boot 1.1.4  (Jul  4 2017)        *
*********************************************

AP121 (AR9331) U-Boot for GL-AR150

DRAM:   64 MB DDR2 16-bit
FLASH:  Winbond W25Q128 (16 MB)
CLOCKS: 400/400/200/33 MHz (CPU/RAM/AHB/SPI)

** Warning: bad env CRC, using default,
   use 'saveenv' to save it in FLASH

LED on during eth initialization...

Hit any key to stop autobooting:  0 

Found ART, checking calibration status...
Device calibrated. Checking MAC address...
Found MAC. Starting firmware...
Booting image at: 0x9F050000

   Image name:   MIPS OpenWrt Linux-5.4.80
   Created:      2020-11-27  13:46:13 UTC
   Image type:   MIPS Linux Kernel Image (lzma compressed)
   Data size:    2130464 Bytes = 2 MB
   Load address: 0x80060000
   Entry point:  0x80060000

Uncompressing kernel image... OK!
Starting kernel...

[    0.000000] Linux version 5.4.80 (builder@buildhost) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r15089-b0cb305236)) #0 Sun Nov 29 15:40:32 2
```

**Comment by ptpt52** [comment]
@1Jeff1  try `sysupgrade -v ...` to see what output in console.

**Comment by s-2** [comment]
Just some wild guessing: could this be related to `gpio_lan_reset` being de-asserted before sysupgrade, and some driver getting confused about the usb-ethernet chip suddenly going to reset etc.? Does this also happen when trying to sysupgrade over wifi?
Besides, I am as well curious about `sysupgrade -v` output, preferably when called directly via uart.

**Comment by ptpt52** [comment]
This is like an incredible issue, I tried my best to debug, but I still can’t locate the cause.
what I get about this:
1.
`/sbin/upgraded` is called

I try even change code in `upgraded.c` not progress.

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts (hunk 4 above)
Comment (981213) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:80]: num-chipselects is documented as a required property for spi-gpio:
https://www.kernel.org/doc/Documentation/devicetree/bindings/spi/spi-gpio.yaml
but it isn't actually used by the spi-gpio driver. I don't know what's going on with this one.
num-chipselects and num-cs under ath79-spi node can be dropped indeed.

**Comment by 1Jeff1** [comment]
> @1Jeff1 try `sysupgrade -v ...` to see what output in console.

I tried sysupgrade -v , it didn't make any difference in the console output, it reboot before start the upgrade.

File: target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts (hunk 4 above)
Comment (adschm) [reply target/linux/ath79/dts/ar9331_glinet_gl-usb150.dts:80]: Thanks for pointing me towards this. I have removed both num-chipselects and num-cs from spi controller node in ath79 in separate patches:
https://patchwork.ozlabs.org/project/openwrt/patch/20201129160717.1607-1-freifunk@adrianschmutzler.de/
https://patchwork.ozlabs.org/project/openwrt/patch/20201129170348.6077-1-freifunk@adrianschmutzler.de/

I have not touched the definitions actually made for spi-gpio so far.

**Comment by ynezz** [comment]
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

**Comment by parkerlreed** [comment]
I know this is a long shot but this might be the easiest place to ask. Does this hardware support switching the single USB port to host instead of enumerating an Ethernet device. 

I want to forgo the USB function and connect a USB device to the router.

**Comment by s-2** [comment]
> Does this hardware support switching the single USB port to host instead of enumerating an Ethernet device?

No, this device does not support USB gadget mode (or at least it would require some soldering). 
As mentioned in the description, it contains a Realtek USB Ethernet adapter (just like those with an RJ-45 port sold as laptop accessories), which is internally connected to one of the ports of the router SoC on the PCB.

**Comment by adschm** [comment]
@ptpt52 @s-2 This is the only device in entire ath79 that does not enable &uart. Is this intentional or should this be fixed?

**Comment by ptpt52** [comment]
It should be fixed.

**Comment by adschm** [comment]
Okay, that's what I thought. I fixed it in master and 21.02.

**Comment by ptpt52** [comment]
great news

I notice that the sysupgrade issue has been solved after this commit https://github.com/openwrt/openwrt/commit/722f1bd5490abafc5d77420905179aa03c508a1f

This issue has troubled me for a long time, and I am puzzled

@adschm @1Jeff1 @s-2

**Comment by xback** [comment]
Fwiw,
I recall having the same reboot issue years ago on imx6 when I used all uarts for other purposes.
As soon as sysupgrade starts it rebooted before upgrade.
I debugged it partially back then to an openwrt system binary which faults out when it doesnt have a valid serial port.

I can maybe check in detail next week which binary it was.

**Comment by 1Jeff1** [comment]
> great news
> 
> I notice that the sysupgrade issue has been solved after this commit [722f1bd](https://github.com/openwrt/openwrt/commit/722f1bd5490abafc5d77420905179aa03c508a1f)
> 
> This issue has troubled me for a long time, and I am puzzled
> 
> @adschm @1Jeff1 @s-2

@ptpt52 I can confirm is fixed.
```
- watchdog -
killall: telnetd: no process killed
Sat Feb 27 13:53:05 UTC 2021 upgrade: Sending TERM to remaining processes ... hostapd wpa_supplicant[ 1226.546943] wlan0: deauthenticating from a1:22:ae:c1:c6:1a by local choice (Reason: 3=DEAUTH_LEAVING)
 netifd odhcpd uhttpd ntpd ttyd dnsmasq sh ubus ubusd urngd logd rpcd
Sat Feb 27 13:53:08 UTC 2021 upgrade: Sending KILL to remaining processes ...
[ 1230.888359] sh (2804): drop_caches: 3
Sat Feb 27 13:53:10 UTC 2021 upgrade: Switching to ramdisk...
mount: mounting /dev/mtdblock5 on /overlay failed: Resource busy
[ 1236.154319] VFS: Busy inodes after unmount of jffs2. Self-destruct in 5 seconds.  Have a nice day...
Sat Feb 27 13:53:15 UTC 2021 upgrade: Performing system upgrade...
[ 1236.259483] sh (2804): drop_caches: 3
Unlocking firmware ...

Writing from <stdin> to firmware ...     
Appending jffs2 data from /tmp/sysupgrade.tgz to firmware..
    
Sat Feb 27 13:53:50 UTC 2021 upgrade: Upgrade completed
Sat Feb 27 13:53:51 UTC 2021 upgrade: Rebooting system...
umount: can't unmount /dev: Resource busy
umount: can't unmount /tmp: Resource busy
[ 1272.265176] reboot: Re��

*********************************************
*        U-Boot 1.1.4  (Jul  4 2017)        *
*********************************************

AP121 (AR9331) U-Boot for GL-AR150

DRAM:   64 MB DDR2 16-bit
FLASH:  Winbond W25Q128 (16 MB)
CLOCKS: 400/400/200/33 MHz (CPU/RAM/AHB/SPI)

** Warning: bad env CRC, using default,
   use 'saveenv' to save it in FLASH

LED on during eth initialization...

Hit any key to stop autobooting:  0 

Found ART, checking calibration status...
Device calibrated. Checking MAC address...
Found MAC. Starting firmware...
Booting image at: 0x9F050000

   Image name:   MIPS OpenWrt Linux-5.10.18
   Created:      2021-02-27   9:14:23 UTC
   Image type:   MIPS Linux Kernel Image (lzma compressed)
   Data size:    2183174 Bytes = 2.1 MB
   Load address: 0x80060000
   Entry point:  0x80060000

Uncompressing kernel image... OK!
Starting kernel...

[    0.000000] Linux version 5.10.18 (jeff@PC) (mips-openwrt-linux-musl-gcc (OpenWrt GCC 8.4.0 r15993-464451d9ab) 8.4.0, GNU ld (GNU Binutils) 2.34) #0 Sat Feb 27 09:14:23 2021
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019374 (MIPS 24Kc)
[    0.000000] MIPS: machine is GL.iNet GL-USB150
[    0.000000] SoC: Atheros AR9330 rev 1
```
